### ISSUE TITLE
clang14 crashes compiling x264 on mips

### ISSUE THREAD

**Issue by kraj** [llvm/llvm-project#53722]
clang13 works ok. Clang14 is crashing on mips/linux works ok when disabling optimizations ( -O0)

```
fatal error: error in backend: Cannot select: 0x5586944fb158: f64 = fabs nnan ninf nsz arcp contract afn reassoc 0x55869449b358, encoder/encoder.c:689:18
  0x55869449b358: f64 = fadd nnan ninf nsz arcp contract afn reassoc 0x558694124be0, 0x55869449b560, encoder/encoder.c:689:47
    0x558694124be0: f64 = fp_extend 0x558694124e50, encoder/encoder.c:689:23
      0x558694124e50: f32,ch = load<(load (s32) from %ir.387, !tbaa !3430)> 0x558694a028a8, 0x558694121a78, undef:i32, encoder/encoder.c:689:35
        0x558694121a78: i32 = add 0x5586940fd730, Constant:i32<684>, encoder/encoder.c:689:35
          0x5586940fd730: i32,ch = CopyFromReg 0x558694a028a8, Register:i32 %176, encoder/encoder.c:682:35
            0x5586944d6710: i32 = Register %176
          0x55869449c260: i32 = Constant<684>
        0x5586944d6c58: i32 = undef
    0x55869449b560: f64,ch = load<(load (s64) from constant-pool)> 0x558694a028a8, 0x558694a3c530, undef:i32
      0x558694a3c530: i32 = add 0x5586944fb430, 0x558693eea8d8
        0x5586944fb430: i32,ch = load<(load (s32) from got)> 0x558694a028a8, 0x558694204840, undef:i32
          0x558694204840: i32 = MipsISD::Wrapper Register:i32 %178, TargetConstantPool:i32<double -1.250000e+00> 0 [TF=1]
            0x55869449b700: i32 = Register %178
            0x5586944fae80: i32 = TargetConstantPool<double -1.250000e+00> 0 [TF=1]
          0x5586944d6c58: i32 = undef
        0x558693eea8d8: i32 = MipsISD::Lo TargetConstantPool:i32<double -1.250000e+00> 0 [TF=5]
          0x558694124ff0: i32 = TargetConstantPool<double -1.250000e+00> 0 [TF=5]
      0x5586944d6c58: i32 = undef
In function: validate_parameters

```

Attached is the testcase to reproduce the segfault.
[encoder-7df1e7.tar.gz](https://github.com/llvm/llvm-project/files/8043638/encoder-7df1e7.tar.gz)

**Comment by kraj** [comment]
another testcase seems similar crash
[steadystate-5a3f60.tar.gz](https://github.com/llvm/llvm-project/files/8126428/steadystate-5a3f60.tar.gz)

**Comment by atanasyan** [comment]
Current state of the investigation.

Minimal reproduction script:
``` llvm
% cat fail.ll
declare double @llvm.fabs.f64(double)

define dso_local double @foo(double %a) #0 {
entry:
  %0 = tail call fast double @llvm.fabs.f64(double %a)
  ret double %0
}

define dso_local double @bar(double %a) {
entry:
  %0 = tail call fast double @llvm.fabs.f64(double %a)
  ret double %0
}

attributes #0 = { nounwind "no-nans-fp-math"="true" }

% llc -march=mips -mcpu=mips32r2 < fail.ll
LLVM ERROR: Cannot select: t3: f64 = fabs nnan ninf nsz arcp contract afn reassoc t2
  t2: f64,ch = CopyFromReg t0, Register:f64 %0
    t1: f64 = Register %0
In function: bar
```

The regression appeared after the 0f0e31c. But I think this change just unhide the another problem. `MipsInstrFPU.td` has a rule for lowering `fabs` to the `abs.d` instruction. It work if either `abs2008` or `no-nans-fp-math` are enabled. Also there is a "custom" lowering code in the `lowerFABS()` routine. It looks like when the `no-nans-fp-math` is disabled (by default) and there is a routine marked by the `"no-nans-fp-math"="true"` attribute, "custom" lowering code is not called.

**Comment by tstellar** [comment]
This issue was not resolved in time for 14.0.0 and will need to be fixed in the next release (14.0.1).  If you disagree with this or you feel there was a mistake, please change the milestone back to 'LLVM 14.0.0 Release' so we can discuss.

**Comment by kraj** [comment]
> This issue was not resolved in time for 14.0.0 and will need to be fixed in the next release (14.0.1). If you disagree with this or you feel there was a mistake, please change the milestone back to 'LLVM 14.0.0 Release' so we can discuss.

this is sad :( a workaround is to use `-O0` or use gcc to compile these components. Less them ideal but not end of world.

**Comment by sdardis** [comment]
This should be resolved by [938ed8ae99f92c19cfb92b76a7e20a72409df84a](https://reviews.llvm.org/rG938ed8ae99f92c19cfb92b76a7e20a72409df84a).

**Comment by kraj** [comment]
Thanks @sdardis for fix. @tstellar can this be backported to 14.x release

**Comment by tstellar** [comment]
Merged: 198626ad43fd
